### PR TITLE
Improve coverage

### LINKED ISSUE #78: Improve coverage on Java-specific things
Test stuff such as private constructors and inherited `Enum` methods. Reason to do this is that it makes the actual coverage number more meaningful and will make it easier to find places where we lack coverage in the future.

### PR DESCRIPTION
**Pull request by jvdb**
Resolves #78.

### CODE REVIEW

File: core/src/test/java/io/parsingdata/metal/ComparisonExpressionSemanticsTest.java [hunk 1]
```diff
@@ -16,36 +16,27 @@
...
+import java.util.Collection;
+
+import static io.parsingdata.metal.Shorthand.*;
+import static io.parsingdata.metal.util.EncodingFactory.enc;
+import static io.parsingdata.metal.util.EnvironmentFactory.stream;
+import static io.parsingdata.metal.util.TokenDefinitions.any;
 
 public class ComparisonExpressionSemanticsTest extends ParameterizedParse {
 
+    public static final Charset ASCII = Charset.forName("ISO646-US");
```
Comment (rdvdijk) [inline core/src/test/java/io/parsingdata/metal/ComparisonExpressionSemanticsTest.java]: Use `StandardCharsets.US_ASCII` here.

> Seven-bit ASCII, a.k.a. ISO646-US, a.k.a. the Basic Latin block of the Unicode character set

File: core/src/test/java/io/parsingdata/metal/util/ClassDefinition.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    // Check that a class is final, has a single private constructor and that all
+    // its declared methods are static.
+    public static void checkUtilityClass(Class<?> c) throws ReflectiveOperationException {
+        assertTrue(Modifier.isFinal(c.getModifiers()));
+        final Constructor<?>[] cons = c.getDeclaredConstructors();
+        assertEquals(1, cons.length);
+        assertTrue(Modifier.isPrivate(cons[0].getModifiers()));
+        cons[0].setAccessible(true);
+        cons[0].newInstance();
+        for (Method m : c.getDeclaredMethods()) {
```
Comment (rdvdijk) [inline core/src/test/java/io/parsingdata/metal/util/ClassDefinition.java]: `final Method m`

File: formats/src/main/java/io/parsingdata/metal/format/Callback.java [hunk 3]
```diff
@@ -60,15 +57,18 @@ public OptionalValue eval(final Value v, final Environment env, final Encoding e
...
-                        } catch (final IOException e) {
-                            return OptionalValue.empty();
+                        inf.setInput(value.getValue());
+                        final byte[] dataReceiver = new byte[512];
+                        final ByteArrayOutputStream out = new ByteArrayOutputStream();
+                        while(!inf.finished()) {
+                            try {
+                                final int processed = inf.inflate(dataReceiver);
+                                out.write(dataReceiver, 0, processed);
+                            } catch (DataFormatException e) {
```
Comment (rdvdijk) [inline formats/src/main/java/io/parsingdata/metal/format/Callback.java]: `final DataFormatException e`

File: formats/src/test/java/io/parsingdata/metal/CallbackTest.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    @Test
+    public void crc32Good() {
+        final OptionalValueList result = crc32(con(0x01020304)).eval(stream(), enc());
+        assertEquals(1, result.size);
+        assertTrue(result.head.isPresent());
+        assertArrayEquals(new byte[] { -74, 60, -5, -51 }, result.head.get().getValue());
+    }
+
+    @Test
+    public void inflateDataFormatError() {
```
Comment (rdvdijk) [inline formats/src/test/java/io/parsingdata/metal/CallbackTest.java]: Add a successful inflate test. I think the following bytes should decompress to `metal`:

```hex
0000000: 789c cb4d 2d49 cc01 0006 4402 14         x..M-I....D..
```

(tip: `$ echo -n "metal" | zlib-flate -compress > test.flate`)

File: core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java [hunk 5]
```diff
@@ -212,4 +215,17 @@ private ParseGraph parseResultGraph(final Environment env, final Token def) {
...
 
+    @Test
+    public void getSubRef() throws IOException {
+        final Token smallSub1 = sub(DEF2, last(ref("value1")));
+        final Token smallSub2 = sub(DEF2, last(ref("value1")));
+        final Token composition = seq(DEF1, smallSub1, smallSub2);
+        final ParseResult result = composition.parse(stream(0), enc());
+        assertTrue(result.succeeded);
+        final ParseItemList refs = ByToken.getAll(result.environment.order, smallSub1);
+        assertEquals(2, refs.size);
```
Comment (rdvdijk) [inline core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java]: Why does `refs` have two items? I would only expect a single `ParseRef`?

File: formats/src/test/java/io/parsingdata/metal/CallbackTest.java (hunk 4 above)
Comment (jvdb) [inline formats/src/test/java/io/parsingdata/metal/CallbackTest.java]: Without the first two and last four bytes :) (That's the nowrap=true for you...)

**Pull request merged**